driver-core/src/main/com/mongodb/internal/connection/LoggingCommandEventSender.java

```diff
@@ -71,12 +79,17 @@ class LoggingCommandEventSender implements CommandEventSender {
     @Override
     public void sendStartedEvent() {
         if (loggingRequired()) {
-            String commandString = redactionRequired ? format("{\"%s\": ...", commandName) : getTruncatedJsonCommand();
-            // TODO: log RequestContext?
-            logger.debug(
-                    format("Sending command '%s' with request id %d to database %s on connection [%s] to server %s",
-                            commandString, message.getId(),
-                            message.getNamespace().getDatabaseName(), description.getConnectionId(), description.getServerAddress()));
+            List<Entry> entries = new ArrayList<>();
+            StringBuilder builder = new StringBuilder("Command \"%s\" started on database %s");
+            entries.add(new Entry("commandName", commandName));
+            entries.add(new Entry("databaseName", message.getNamespace().getDatabaseName()));
+
+            appendCommonLogFragment(entries, builder);
+
+            builder.append(" Command: %s");
+            entries.add(new Entry("command", redactionRequired ? "{}" : getTruncatedJsonCommand(commandDocument)));
+
+            logger.log(new StructuredLogMessage(COMMAND, DEBUG, "Command started", getClusterId(), entries), builder.toString());
```

Comment:
[optional]

This is much better that what was previously. However, this code screams "use a specialized builder to make sure that when the builder for the format of an unstructured message is updated, the entries for the structured log message are also updated". So I envision something like

```java
static final class StructuredLogMessageBuilder {
    private final StringBuilder unstructuredMessageFormatBuilder;
    private final List<Entry> entries;
    StructuredLogMessageBuilder() {
        unstructuredMessageFormatBuilder = new StringBuilder();
        entries = new ArrayList<>();
    }
    StructuredLogMessageBuilder append(final String format, final Entry entry) {
        unstructuredMessageFormatBuilder.append(format);
        entries.add(entry);
        return this;
    }
    // These params can be moved to the constructor of `StructuredLogMessageBuilder`
    // and stored as fields there, but for the sake of imaginary efficiency,
    // I did not do that.
    StructuredLogMessage build(
            final StructuredLogMessage.Component component,
            final StructuredLogMessage.Level level,
            final String messageId,
            final ClusterId clusterId,
            @Nullable final Throwable exception) {
        return new StructuredLogMessage(component, level, messageId, clusterId, excep
    }
    String unstructuredMessageFormat() {
        return unstructuredMessageFormatBuilder.toString();
    }
}
```
which results in

```java
StructuredLogMessageBuilder logMessageBuilder = new StructuredLogMessageBuilder()
        .append("Command \"%s\" failed", new Entry("commandName", commandName))
        .append(" in %.2f ms", new Entry("durationMS", elapsedTimeNanos / NANOS_PER_MILLI));
appendCommonLogFragment(logMessageBuilder);
logger.log(
        logMessageBuilder.build(COMMAND, DEBUG, "Command failed", getClusterId(), commandEventException),
        logMessageBuilder.unstructuredMessageFormat());
```

Reply:
Just for information (this thread was and continues to be optional):

1. To achieve _"could easily append a format string fragment that doesn't include exactly one placeholder"_ we may have `append(String format, Entry... entries)` instead of `append(String format, Entry entry)`.
2. We also don't have to use `StructuredLogMessageBuilder` when it's simpler to construct `StructuredLogMessage` without it.